File: src/core/inventory/order.ts
import type { PartRollup } from "../consolidate/findings.js";
import type { CatalogPart, Resolution } from "./types.js";

/**
 * One row of the sheet you fill in.
 *
 * Everything to the left of `supplier` is the tool describing the board;
 * `mpn`, `supplier` and `order_number` are yours. That division is the whole
 * design: the tool counts, you decide.
 */
export type OrderRow = {
  key: string;
  value: string;
  package: string;
  used: number;
  mpn: string;
  manufacturer: string;
  supplier: string;
  order_number: string;
  /** The package of the part *you* buy — left blank on purpose; see CatalogPart. */
  part_package: string;
  refs: string;
};

export const orderColumns: readonly (keyof OrderRow)[] = [
  "key",
  "value",
  "package",
  "used",
  "mpn",
  "manufacturer",
  "supplier",
  "order_number",
  "part_package",
  "refs",
];

/** The board's parts, with whatever the catalog already knows filled in. */
export function orderRows(
  parts: readonly PartRollup[],
  resolutions: readonly Resolution[],
): OrderRow[] {
  const byKey = new Map(resolutions.map((r) => [r.key, r]));
  return parts.map((part) => {
    const resolution = byKey.get(part.key);
    const catalogued = resolution?.part;
    return {
      key: part.key,
      value: part.value,
      package: part.pkg,
      used: part.placements,
      // A part whose value is already a part number arrives with it suggested,
      // the same way the resolve tab offers it — you still have to keep it.
      mpn: catalogued?.mpn ?? resolution?.suggestedMpn ?? "",
      manufacturer: catalogued?.manufacturer ?? "",
      supplier: catalogued?.supplier ?? "",
      order_number: catalogued?.orderNumber ?? "",
      part_package: catalogued?.package ?? "",
      refs: part.refs.join(" "),
    };
  });
}

export type ImportedRow = {
  key: string;
  mpn: string;
  manufacturer: string;
  supplier: string;
  orderNumber: string;
  package: string;
};

export type ImportProblem = { row: number; key: string; reason: string };

/**
 * Reads the sheet back.
 *
 * Rows are matched on the canonical key, which is the one column you are not
 * asked to touch. A key the board does not have is reported rather than
 * guessed at: a spreadsheet that has drifted from the board — an old export, a
 * sorted-and-mangled column — must not quietly write part numbers into the
 * catalog under names nothing will ever look up.
 */
export function readOrderRows(
  records: readonly Record<string, string>[],
  known: ReadonlySet<string>,
): { rows: ImportedRow[]; problems: ImportProblem[] } {
  const rows: ImportedRow[] = [];
  const problems: ImportProblem[] = [];

  records.forEach((record, index) => {
    const line = index + 2; // 1-based, and the header is line 1
    const cell = (name: string): string => (record[name] ?? "").trim();
    const key = cell("key");
    if (key === "") return;

    if (!known.has(key)) {
      problems.push({ row: line, key, reason: "no part on this board has this key" });
      return;
    }

    const mpn = cell("mpn");
    const supplier = cell("supplier");
    const orderNumber = cell("order_number");
    const pkg = cell("part_package");
    if (mpn === "" && supplier === "" && orderNumber === "" && pkg === "") return; // untouched

    if (mpn === "") {
      problems.push({ row: line, key, reason: "a supplier was given but no MPN" });
      return;
    }
    rows.push({
      key,
      mpn,
      manufacturer: cell("manufacturer"),
      supplier,
      orderNumber,
      package: cell("part_package"),
    });
  });

  return { rows, problems };
}

export type OrderLine = {
  key: string;
  quantity: number;
  part: CatalogPart;
  refs: string[];
};

export type SupplierOrder = {
  supplier: string;
  lines: OrderLine[];
};

export type OrderPlan = {
  boards: number;
  orders: SupplierOrder[];
  /** Parts with a part number but nobody to buy them from. */
  noSupplier: OrderLine[];
  /** Parts with no MPN at all — step 2 has not been done for these. */
  unresolved: PartRollup[];
};

/**
 * What to buy, grouped by who you buy it from.
 *
 * The quantity is placements × boards and nothing else. No stock subtraction,
 * no spares, no rounding to a pack size, no price break — every one of those
 * is a decision, and a tool that makes them quietly is a tool that can order
 * ten thousand of a five-euro part because the arithmetic said so.
 */
export function planOrder(
  parts: readonly PartRollup[],
  resolutions: readonly Resolution[],
  boards = 1,
): OrderPlan {
  const byKey = new Map(resolutions.map((r) => [r.key, r]));
  const bySupplier = new Map<string, OrderLine[]>();
  const noSupplier: OrderLine[] = [];
  const unresolved: PartRollup[] = [];

  for (const part of parts) {
    const catalogued = byKey.get(part.key)?.part;
    if (catalogued === undefined) {
      unresolved.push(part);
      continue;
    }
    const line: OrderLine = {
      key: part.key,
      quantity: part.placements * boards,
      part: catalogued,
      refs: part.refs,
    };
    const supplier = catalogued.supplier;
    if (supplier === undefined || supplier === "") {
      noSupplier.push(line);
      continue;
    }
    const lines = bySupplier.get(supplier);
    if (lines === undefined) bySupplier.set(supplier, [line]);
    else lines.push(line);
  }

  const orders = [...bySupplier]
    .map(([supplier, lines]) => ({
      supplier,
      lines: lines.sort((a, b) => a.key.localeCompare(b.key)),
    }))
    .sort((a, b) => a.supplier.localeCompare(b.supplier));

  return { boards, orders, noSupplier, unresolved };
}

/** A filename per supplier, with nothing in it a filesystem will object to. */
export function supplierFileName(supplier: string): string {
  const safe = supplier
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-|-$/g, "");
  return `order.${safe === "" ? "supplier" : safe}.csv`;
}

File: kinv/core/inventory/order.py
"""Step 3: what to buy, grouped by who you buy it from."""

from __future__ import annotations

import re
from typing import Iterable

from kinv.js import collation_key

# One row of the sheet you fill in. Everything left of `supplier` is the tool
# describing the board; `mpn`, `supplier` and `order_number` are yours.
ORDER_COLUMNS = ["key", "value", "package", "used", "mpn", "manufacturer", "supplier", "order_number", "part_package", "refs"]


def order_rows(parts: Iterable[dict], resolutions: Iterable[dict]) -> list[dict]:
    """The board's parts, with whatever the catalog already knows filled in."""
    by_key = {r["key"]: r for r in resolutions}
    rows = []
    for part in parts:
        resolution = by_key.get(part["key"])
        catalogued = resolution["part"] if resolution else None

        def known(field: str) -> str:
            value = catalogued.get(field) if catalogued else None
            return value if value is not None else ""

        suggested = resolution["suggestedMpn"] if resolution else None
        mpn = catalogued["mpn"] if catalogued and catalogued.get("mpn") is not None else (suggested if suggested is not None else "")
        rows.append(
            {
                "key": part["key"],
                "value": part["value"],
                "package": part["pkg"],
                "used": part["placements"],
                # a part whose value is already a part number arrives with it suggested
                "mpn": mpn,
                "manufacturer": known("manufacturer"),
                "supplier": known("supplier"),
                "order_number": known("orderNumber"),
                "part_package": known("package"),
                "refs": " ".join(part["refs"]),
            }
        )
    return rows


def read_order_rows(records: list[dict], known: set[str]) -> dict:
    """Reads the sheet back, matched on the canonical key; a key the board lacks is reported."""
    rows: list[dict] = []
    problems: list[dict] = []
    for index, record in enumerate(records):
        line = index + 2  # 1-based, and the header is line 1

        def cell(name: str) -> str:
            return (record.get(name) or "").strip()

        key = cell("key")
        if key == "":
            continue
        if key not in known:
            problems.append({"row": line, "key": key, "reason": "no part on this board has this key"})
            continue

        mpn, supplier, order_number, pkg = cell("mpn"), cell("supplier"), cell("order_number"), cell("part_package")
        if mpn == "" and supplier == "" and order_number == "" and pkg == "":
            continue  # untouched
        if mpn == "":
            problems.append({"row": line, "key": key, "reason": "a supplier was given but no MPN"})
            continue
        rows.append(
            {
                "key": key,
                "mpn": mpn,
                "manufacturer": cell("manufacturer"),
                "supplier": supplier,
                "orderNumber": order_number,
                "package": cell("part_package"),
            }
        )
    return {"rows": rows, "problems": problems}


def plan_order(parts: Iterable[dict], resolutions: Iterable[dict], boards: int = 1) -> dict:
    """The quantity is placements × boards and nothing else.

    No stock subtraction, no spares, no rounding to a pack, no price break —
    every one is a decision, and a tool that makes them quietly can order ten
    thousand of a five-euro part because the arithmetic said so.
    """
    by_key = {r["key"]: r for r in resolutions}
    by_supplier: dict[str, list[dict]] = {}
    no_supplier: list[dict] = []
    unresolved: list[dict] = []

    for part in parts:
        resolution = by_key.get(part["key"])
        catalogued = resolution["part"] if resolution else None
        if catalogued is None:
            unresolved.append(part)
            continue
        line = {"key": part["key"], "quantity": part["placements"] * boards, "part": catalogued, "refs": part["refs"]}
        supplier = catalogued.get("supplier")
        if not supplier:
            no_supplier.append(line)
            continue
        by_supplier.setdefault(supplier, []).append(line)

    orders = [
        {"supplier": supplier, "lines": sorted(lines, key=lambda l: collation_key(l["key"]))}
        for supplier, lines in by_supplier.items()
    ]
    orders.sort(key=lambda o: collation_key(o["supplier"]))
    return {"boards": boards, "orders": orders, "noSupplier": no_supplier, "unresolved": unresolved}


def supplier_file_name(supplier: str) -> str:
    """A filename per supplier, with nothing in it a filesystem will object to."""
    safe = re.sub(r"^-|-\Z", "", re.sub(r"[^a-z0-9]+", "-", supplier.lower()))
    return f"order.{safe or 'supplier'}.csv"
